packages/slate/src/classes/commands/mark.ts

```diff
@@ -0,0 +1,159 @@
+      hanging?: boolean
+    } = {}
+  ) {
+    this.withoutNormalizing(() => {
+      const at = splitLocation(this, options)
+
+      if (!at) {
+        return
+      }
+
+      // De-dupe the marks being added to ensure the set is unique.
+      const set: Mark[] = []
+
+      for (const mark of marks) {
+        if (!Mark.exists(mark, set)) {
+          set.push(mark)
+        }
+      }
+
+      for (const [node, path] of this.texts({ at })) {
+        for (const mark of set) {
+          if (!Mark.exists(mark, node.marks)) {
+            this.apply({ type: 'add_mark', path, mark })
+          }
+        }
+      }
+    })
+  }
+
+  removeMarks(
+    this: Editor,
+    marks: Mark[],
+    options: {
+      at?: Location
+      hanging?: boolean
+    } = {}
+  ) {
+    this.withoutNormalizing(() => {
+      const at = splitLocation(this, options)
+
+      if (at) {
+        for (const [mark, i, node, path] of this.marks({ at })) {
+          if (Mark.exists(mark, marks)) {
+            this.apply({ type: 'remove_mark', path, mark })
+          }
+        }
+      }
+    })
+  }
+
+  setMarks(
+    this: Editor,
+    marks: Mark[],
+    props: Partial<Mark>,
+    options: {
+      at?: Location
+      hanging?: boolean
+    } = {}
+  ) {
+    this.withoutNormalizing(() => {
+      const at = splitLocation(this, options)
+
+      if (at) {
+        for (const [mark, i, node, path] of this.marks({ at })) {
+          if (Mark.exists(mark, marks)) {
+            const newProps = {}
+
+            for (const k in props) {
+              if (props[k] !== mark[k]) {
+                newProps[k] = props[k]
+              }
+            }
+
+            if (Object.keys(newProps).length > 0) {
+              this.apply({
+                type: 'set_mark',
+                path,
+                properties: mark,
+                newProperties: newProps,
+              })
+            }
+          }
+        }
+      }
+    })
+  }
+
+  toggleMarks(
+    this: Editor,
+    marks: Mark[],
+    options: {
+      at?: Location
+      hanging?: boolean
+    } = {}
+  ) {
+    this.withoutNormalizing(() => {
+      const existing = this.getActiveMarks(options)
+      const exists = marks.every(m => Mark.exists(m, existing))
+
+      if (exists) {
+        this.removeMarks(marks, options)
+      } else {
+        this.addMarks(marks, options)
+      }
+    })
+  }
+}
+
+/**
+ * Split the text nodes at a range's edges to prepare for adding/removing marks.
+ */
+
+const splitLocation = (
+  editor: Editor,
+  options: {
+    at?: Location
+    hanging?: boolean
+  } = {}
+): Location | undefined => {
+  let { at = editor.value.selection, hanging = false } = options
+
+  if (!at) {
+    return
+  }
+
+  if (Range.isRange(at)) {
+    if (!hanging) {
+      at = editor.unhangRange(at)
+    }
+
+    const rangeRef = editor.createRangeRef(at, { affinity: 'inward' })
+    const [start, end] = Range.edges(at)
+    editor.splitNodes({ at: end, match: 'text' })
```

Comment:
Do one/both of these `splitNodes` need `always: true`. This function does not seem to be able to apply a mark to a collapsed selection.

Reply:
When you say to a collapsed selection, what are you expecting for the output?

For the case of "future marks", where you hit <kbd>cmd+b</kbd> and then type some text in bold, I'm thinking having those being a `slate-react`-only concern that is stored as local state. That way we don't need operations. Similar to how `editor.isFocused()` is now a DOM-only concern instead of in core.

Reply:
It's the future marks case I'm talking about like you describe (cmd+b at the end of a node to make future text bold). Whatever works I think is fine, in the future it should just make its way to the rich-text editor example which had that as part of it (to keep it functionally equivalent to the current version).